Cpp wrappers add version safety

jcohen-nvidia opened the pull request:
Add version-safety to nvtx3.hpp

This change does not affect the existing interface.  It introduces the concepts "versioned symbols" and "unversioned symbols", where the latter is the familiar interface like nvtx3::mark and NVTX3_FUNC_RANGE.  The corresponding versioned symbols would be nvtx3::v1::mark and NVTX3_V1_FUNC_RANGE.  Versioned symbols are always provided, and unversioned symbols are provided by default, but defining NVTX3_CPP_REQUIRE_EXPLICIT_VERSION allows library developers to opt-out of the unversioned symbols.  This is important for adding instrumentation in header-based libraries, where a single .cpp file might include multiple header-only libraries that in turn include different versions of nvtx3.hpp.  By defining NVTX3_CPP_REQUIRE_EXPLICIT_VERSION and only using the versioned symbols, a header-only library guarantees that it will not create conflicts with any other NVTX usage in other code.

This change also introduces major and minor versions following the rules of semantic versioning.  Major version number changes imply breaking changes, so clients may need to change their code after updating the library to a newer major version.  Minor version number changes imply purely additive changes, so updating to a new minor version within the same major version guarantees no client changes will be needed.  Also, multiple minor versions may be used within the same translation unit, and functionality accumulates -- order of include directives is irrelevant.

Up to one major version can provide unversioned symbols, including any number of minor versions within it.  A clear error is emitted if the client tries to use two different major versions with unversioned symbols.

Testing this change was difficult, since there's only one version so far. I synthesized fake version 1.1 and 2.0 copies of nvtx3.hpp and verified that all the aforementioned behavior works as expected with a variety of app sets. The three major categories of tests are:
- Compiler level: two versions included into the same .cpp file
- Linker level: two versions compiled into separate .cpp files & linked
- DLL level: two versions compiled/linked into separate DLLs & loaded

jcohen-nvidia committed:
Fix one last usage of "and" instead of "&&"

We shouldn't use these since MSVC doesn't support them unless you #include an
extra header, which can cause problems in some codebases.

jcohen-nvidia committed:
Remove a stray comment ("namespace nvtx3" on end of domain class def)

jcohen-nvidia committed:
Change "nvtx3::domain" to just "domain" in mark()

Mark is being defined inside namespace nvtx3, so the nvtx3:: is not needed,
and using it causes problems with the nested version-specific namespace when
not inlining that namespace.

jcohen-nvidia committed:
Add version-safety to nvtx3.hpp

This change does not affect the existing interface.  It introduces the concepts
"versioned symbols" and "unversioned symbols", where the latter is the familiar
interface like nvtx3::mark and NVTX3_FUNC_RANGE.  The corresponding versioned
symbols would be nvtx3::v1::mark and NVTX3_V1_FUNC_RANGE.  Versioned symbols are
always provided, and unversioned symbols are provided by default, but defining
NVTX3_CPP_REQUIRE_EXPLICIT_VERSION allows library developers to opt-out of the
unversioned symbols.  This is important for adding instrumentation in header-
based libraries, where a single .cpp file might include multiple header-only
libraries that in turn include different versions of nvtx3.hpp.  By defining
NVTX3_CPP_REQUIRE_EXPLICIT_VERSION and only using the versioned symbols, a
header-only library guarantees that it will not create conflicts with any other
NVTX usage in other code.

This change also introduces major and minor versions following the rules of
semantic versioning.  Major version number changes imply breaking changes, so
clients may need to change their code after updating the library to a newer
major version.  Minor version number changes imply purely additive changes,
so updating to a new minor version within the same major version guarantees
no client changes will be needed.  Also, multiple minor versions may be used
within the same translation unit, and functionality accumulates -- order of
include directives is irrelevant.

Up to one major version can provide unversioned symbols, including any number
of minor versions within it.  A clear error is emitted if the client tries to
use two different major versions with unversioned symbols.

Testing this change was difficult, since there's only one version so far. I
synthesized fake version 1.1 and 2.0 copies of nvtx3.hpp and verified that
all the aforementioned behavior works as expected with a variety of app sets.
The three major categories of tests are:
- Compiler level: two versions included into the same .cpp file
- Linker level: two versions compiled into separate .cpp files & linked
- DLL level: two versions compiled/linked into separate DLLs & loaded

jcohen-nvidia commented:
I noticed there's a clang-format being applied, and that made it tricky to format the ugly nested macro defines at the top of the file nicely.  Here's how I had that section in my test code:

```
// Inline unversioned symbols, unless NVTX3_CPP_REQUIRE_EXPLICIT_VERSION is defined
#if !defined(NVTX3_CPP_REQUIRE_EXPLICIT_VERSION)
    #define NVTX3_INLINE_THIS_VERSION
    #if !defined(NVTX3_CPP_INLINED_VERSION_MAJOR)
        // This is the first occurrence of this header in the translation unit
        #define NVTX3_CPP_INLINED_VERSION_MAJOR 1 //NVTX3_CPP_VERSION_MAJOR
        #define NVTX3_CPP_INLINED_VERSION_MINOR 0 //NVTX3_CPP_VERSION_MINOR
    #elif NVTX3_CPP_INLINED_VERSION_MAJOR != NVTX3_CPP_VERSION_MAJOR
        #error "Two different major versions of the NVTX C++ Wrappers are being included in a single .cpp file, with unversioned symbols enabled.  Only one major version can enable unversioned symbols in a .cpp file.  To disable unversioned symbols, #define NVTX3_CPP_REQUIRE_EXPLICIT_VERSION before #including nvtx3.hpp, and use the explicit-version symbols instead -- this is the preferred way to use nvtx3.hpp from a header file."
    #elif (NVTX3_CPP_INLINED_VERSION_MAJOR == NVTX3_CPP_VERSION_MAJOR) && (NVTX3_CPP_INLINED_VERSION_MINOR < NVTX3_CPP_VERSION_MINOR)
        // Already inlined an older minor version, so bump up to the current minor
        // version, since its features will be added to the inlining.
        #undef NVTX3_CPP_INLINED_VERSION_MINOR
        #define NVTX3_CPP_INLINED_VERSION_MINOR 0 //NVTX3_CPP_VERSION_MINOR
    // else, already have this version or newer, nothing to do
    #endif
#endif
```

Would you guys prefer we override the clang-format and do something like this to improve legibility?

jcohen-nvidia commented:
Also, proactively assuming someone will comment on this (I would! haha)...   I had to do things like this a few times:
```
        #define NVTX3_CPP_INLINED_VERSION_MAJOR 1 //NVTX3_CPP_VERSION_MAJOR
        #define NVTX3_CPP_INLINED_VERSION_MINOR 0 //NVTX3_CPP_VERSION_MINOR
```
redundantly hard-coding the value instead of instead of the more obvious
```
        #define NVTX3_CPP_INLINED_VERSION_MAJOR NVTX3_CPP_VERSION_MAJOR
        #define NVTX3_CPP_INLINED_VERSION_MINOR NVTX3_CPP_VERSION_MINOR
```
Unfortunately defining a macro to a macro made it not resolve properly in some cases.  I figure we can fix it properly somehow and avoid having more places to update when changing those numbers, but there's no reason to delay this for a nitpick like that, given how late it is already.  We can ensure any fixes like that in the future play nicely with the prior versions.

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -15,21 +15,67 @@
-#endif
-
+/* Temporary helper #defines, #undef'ed at end of header */
+#define NVTX3_CPP_VERSION_MAJOR 1
```
Do we want to keep the `NVTX3` name if we're starting the major version off at 1? These seem like a potential source of confusion if they're not consistent. Instead we could just go with `NVTX`? If not, do we intend to change to `nvtx4` when that comes out?

There's a few different moving pieces wrt to versioning between NVTX C and NVTX C++ and we should try and keep them as consistent as possible.

dhoro-nvidia replied:
You could follow what the NVTX extensions are doing with calling it compatID to version the extension separately but not make people as confused over all the versions.  CompatID lets me pick marketing hubbub totally separate from code.  Yes v4 likely could warrant its own set of C++ wrappers that can live SxS with v3 C++ wrapper.  This way people can have v3 wrappers optimal for v3 and support v3 compatible tools while v4 requires upgraded tools and could have more efficient ways to do the same thing not possible with v3.

jcohen-nvidia replied:
I think v4 will end up looking quite a bit different from v3, so I like keeping the C++ wrappers for v3 "scoped" to that version.  It certainly keeps us safe from conflicts in the future.  Even for these temporary macros which are unversioned with respect to the C++ wrappers (i.e. no "_V1" in the macro name) because I just #undef them at the end of the header, if I changed e.g. "NVTX3_CONCAT" to "NVTX_CONCAT", we'd have to be careful in the future of both C and C++ NVTX impls to not use that macro name, even temporarily.  It's annoying that you can't "push" and "pop" #defines, so they don't conflict with pre-existing definitions.  So just being careful to namespace them is a good seatbelt there.

For v4 I am fairly convinced we shouldn't have C++ wrappers around a C header impl, and just write the C and C++ versions directly to the ABI, separately.  I think that will also reduce a lot of the confusion as you say about the moving pieces between core impl and wrappers.

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -15,21 +15,67 @@
+#define NVTX3_CPP_VERSION_MINOR 0
+
+/* This section handles the decision of whether to provide unversioned symbols.
+ * If NVTX3_CPP_REQUIRE_EXPLICIT_VERSION is #defined, unversioned symbols are
```
An example here showing good/bad file include orders would help someone understand what this is all for.

jcohen-nvidia replied:
Would you be OK with me adding that more detailed info to the DITA docs instead of here in the comments?  I think it's tricky to explain well without diagrams and pretty formatting, and I also don't like making these files get humongous just with comments.  I already have a versioning guide for NVTX and intend to do one for NVTX++ too.

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -15,21 +15,67 @@
+ * to coexist without problems within a translation unit, the recommended best
+ * practice for instrumenting header-based libraries with NVTX C++ Wrappers is
+ * is to #define NVTX3_CPP_REQUIRE_EXPLICIT_VERSION before including nvtx3.hpp,
+ * #undef it afterward, and only use explicit-version symbols.  This is not
```
Example here would be good as well.

jcohen-nvidia replied:
ditto on above comments

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -15,21 +15,67 @@
+ */
+#if !defined(NVTX3_CPP_REQUIRE_EXPLICIT_VERSION)
+#define NVTX3_INLINE_THIS_VERSION
+#if !defined(NVTX3_CPP_INLINED_VERSION_MAJOR)
+// First occurrence of this header in the translation unit
+/**
+ * @brief Semantic major version number for NVTX C++ wrappers of unversioned symbols
+ *
+ * Breaking changes may occur between major versions, and different major versions
+ * cannot provide unversioned symbols in the same translation unit (.cpp file).
+ * 
+ * Note: If NVTX3_CPP_REQUIRE_EXPLICIT_VERSION is defined, this macro is not defined.
+ *
+ * Not to be confused with the version number of the NVTX core library.
+ */
+#define NVTX3_CPP_INLINED_VERSION_MAJOR 1  // NVTX3_CPP_VERSION_MAJOR
```
Separating the `#define` from the `if` with a comment block made it hard to see what was inside the `if`. 
```suggestion
/**
 * @brief Semantic major/minor version number for NVTX C++ wrappers of unversioned symbols
 *
 * Breaking changes may occur between major versions, and different major versions
 * cannot provide unversioned symbols in the same translation unit (.cpp file).
 * 
 * No breaking changes occur between minor versions -- minor version changes within
 * a major version are purely additive.
 * 
 * Not to be confused with the version number of the NVTX core library.
 *
 * @note: If NVTX3_CPP_REQUIRE_EXPLICIT_VERSION is defined, this macro is not defined.
 *
 */
#if !defined(NVTX3_CPP_INLINED_VERSION_MAJOR) // First occurrence of this header in the translation unit
#define NVTX3_CPP_INLINED_VERSION_MAJOR 1  // NVTX3_CPP_VERSION_MAJOR
#define NVTX3_CPP_INLINED_VERSION_MINOR 0 // NVTX3_CPP_VERSION_MINOR
```

jcohen-nvidia replied:
I didn't really like having the doxygen bloating the inside of this macro logic either, but is there a way to do what you're suggesting (moving the doxygen comments for NVTX3_CPP_INLINED_VERSION_MAJOR  & _MINOR outside the #if) while still connecting the comments to those macros?  Like, I want to make sure those macros in particular are easy to find the docs for...   in IDEs that show doxygen in tooltips, when searching the doxygen pages for them, etc.  I'm happy to make this change if we can keep that relationship -- I just don't know doxygen enough to make that work.

dhoro-nvidia replied:
This won't affect Katie and others right?  Maybe we can fix this later?  Maybe you can use a doxygen sub-pages to cover this topic and shorten the doxygen here to mention reading this sub-page for more details?

jcohen-nvidia replied:
Yeah, in my amend on this comment I am turning off clang format for this chunk, the indenting makes it way easier to read.  While I'd like to move these comments out and refer to these macros with a doxygen reference from somewhere else, I couldn't figure out how to do that yet, so I'm happy leaving it this way for now and fixing that later.  There are no version complications at all from just fixing comments whenever we want.  And the nice thing about the current approach is that we get proper doxygen association for these macros by putting the comments right above them.  (I ran doxygen and checked)

jcohen-nvidia replied:
OK I fixed the indenting so this is all way more readable.

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -2124,4 +2189,54 @@ inline void mark(event_attributes const& attr) noexcept
+ * that the versioned symbols and macros are always provided, and the unversioned symbols
+ * are only provided if NVTX3_INLINE_THIS_VERSION was defined at the top of this header.
+ *
+ * #ifndef NVTX3_CPP_DEFINITIONS_V1_2
```
```suggestion
 * \code{.cpp}
 * #ifndef NVTX3_CPP_DEFINITIONS_V1_2
```

jcohen-nvidia replied:
Oh cool, I will do that.

jcohen-nvidia replied:
Done

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -2124,4 +2189,54 @@ inline void mark(event_attributes const& attr) noexcept
+ *         #define NVTX3_NEW_MACRO_A NVTX3_V1_NEW_MACRO_A
+ *         #define NVTX3_NEW_MACRO_B NVTX3_V1_NEW_MACRO_B
+ *     #endif
+ * #endif // NVTX3_CPP_DEFINITIONS_V1_2
```
```suggestion
 * #endif // NVTX3_CPP_DEFINITIONS_V1_2
 * \endcode
```

jcohen-nvidia replied:
ditto

jcohen-nvidia replied:
done

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -502,22 +548,39 @@
+ * where "X" is the major version number. */
+#define NVTX3_CONCAT(A, B) A##B
+#define NVTX3_NAMESPACE_FOR(VERSION) NVTX3_CONCAT(v, VERSION)
+#define NVTX3_VERSION_NAMESPACE NVTX3_NAMESPACE_FOR(NVTX3_CPP_VERSION_MAJOR)
```
Seems cleaner to put the `inline` check here instead of below where the namespace is actually opened. 
```suggestion
#define NVTX3_VERSION_NAMESPACE NVTX3_NAMESPACE_FOR(NVTX3_CPP_VERSION_MAJOR)
#if defined(NVTX3_INLINE_THIS_VERSION)
#define NVTX3_VERSIONED_NAMESPACE inline NVTX3_VERSION_NAMESPACE
#else
#define NVTX3_VERSIONED_NAMESPACE NVTX3_VERSION_NAMESPACE
#endif
```

jcohen-nvidia replied:
Interesting idea.  I was just following pattern I saw in a few places, like here:

https://stackoverflow.com/questions/11016220/what-are-inline-namespaces-for

I'd have to do it a little differently than you suggested because your macros here resolve to `namespace inline v1` and it needs to be `inline namespace v1`.  So we could do
```
#if defined(NVTX3_INLINE_THIS_VERSION)
#define NVTX3_INLINE_IF_REQUESTED inline
#else
#define NVTX3_INLINE_IF_REQUESTED
#endif
```
 then the usage below would look like `NVTX3_INLINE_IF_REQUESTED namespace NVTX3_VERSION_NAMESPACE {`.  Do you like that better than the way I currently have it?

jcohen-nvidia replied:
Note also that the `#if defined(NVTX3_INLINE_THIS_VERSION)` thing is still needed below in each minor version where it defines the NVTX3_WHATEVER macros as aliases of NVTX3_V1_WHATEVER, so this change doesn't let us be totally free of that pattern.

jcohen-nvidia replied:
OK I made this change and I liked it better than what I had.  Fixed in both the V1_0 impl and the example code in comments describing how to add new functionality in minor versions.

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -502,22 +548,39 @@
+
 namespace nvtx3 {
+
+#if defined(NVTX3_INLINE_THIS_VERSION)
+inline
+#endif
+namespace NVTX3_VERSION_NAMESPACE {
```
Assuming changes from above.
```suggestion
namespace NVTX3_VERIONSED_NAMESPACE {
```

jcohen-nvidia replied:
See comments in above blurb -- resolving this one

jrhemstad commented on cpp/nvtx3.hpp:
```diff
@@ -15,21 +15,67 @@
  */
 #pragma once
```
Probably should get rid of this `pramga once` is it will likely interfere with all the custom include guard stuff.

jcohen-nvidia replied:
Yeah I was thinking about this, and I realized there is a case where it's a problem...   If someone includes nvtx3.hpp using the #define to disable the inlining, and then later someone else includes the exact same header without that define & uses the unversioned symbols, they'd be broken.  So yeah, I'll take out the #pragma once.

jcohen-nvidia replied:
done

jcohen-nvidia committed:
Enable MACRO_EXPANSION in Doxygen configuration for C++ wrappers

With the new versioning scheme, the inline versioned namespace has a convenience
macro NVTX_VERSION_NAMESPACE which resolves to v1.  This macro is internal to
the header and is undef'ed at the end.  With MACRO_EXPANSION off, Doxygen shows
class names like "nvtx3::NVTX_VERSION_NAMESPACE::domain" instead of the way it
would show up to users, i.e. "nvtx3::v1::domain".  Enabling macro expansion
fixes the Doxygen output to show the resolved version.  This does not appear to
make any other documentation incorrect.

jcohen-nvidia commented:
I added a commit to enable macro expansion in the Doxygen config.  It makes the name of the inline namespace appear the way the user will see it, and not as the temporary macro which is only defined within the nvtx3.hpp header.
